### PR TITLE
Clear cache api

### PR DESCRIPTION
**Pull request by replay**
Adds an API endpoint to clear the cache by metric name, including all the chunks of associated aggregations.

https://github.com/raintank/metrictank/issues/547

### CODE REVIEW

**Comment by woodsaj** [comment]
Does this api endpoint need to broadcast the request to all nodes in the cluster?  I feel that it should, or at least should have a flag to tell it to.

**Comment by replay** [comment]
Interesting idea. What's the reason why you think that's necessary, because the cluster also forwards the requests between the nodes if something requested is known to be another node? That makes sense

**Comment by woodsaj** [comment]
If running a multinode cluster, and i want to delete the cache for all series that match "some.metric.*" the series that match that pattern could be spread across many nodes.  

Having the delete request propagate through the whole cluster would be a requirement for exposing this feature to users who would typically have no visibility into the cluster tolopolgy, they only have a single gateway address.

**Comment by Dieterbe** [comment]
+1 to support propagation. but we should make it optional. I suspect often enough cache clearing will be done by an operator who wants to only do it to 1 instance, to diagnose or try to fix a problem, or the operator may want to do it to different instances at different times, instead of all at once.

**Comment by replay** [comment]
@Dieterbe  @woodsaj  took a while, but this would be ready to review again

**Comment by Dieterbe** [comment]
it seems to make sense to be able to delete the entire cache, currently it can only be done for series matching the pattern, but there is no pattern to mach all series (correct me if i'm wrong), so maybe allow for pattern `""` or `"*"` and delete entire cache in that case?

File: api/ccache.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		}
+		for _, node := range nodes {
+			for _, def := range node.Defs {
+				delResult := s.Cache.DelMetric(def.Id)
+				res.DeletedSeries += delResult.Series
+				res.DeletedArchives += delResult.Archives
+			}
+		}
+	}
+	response.Write(ctx, response.NewJson(200, res, ""))
```
Comment (Dieterbe) [inline api/ccache.go]: if the peers reported a bunch of errors, or could not connect to peers, we should probably not report 200 ok

**Comment by Dieterbe** [comment]
seems like something went wrong with the rebase. it's showing a bunch of changes that seem unrelated to this PR, like moving cluster stuff around. can you amend the commit so that the diff is only what is supposed to change

**Comment by replay** [comment]
I think you're right that it would make sense to add a way to clear the whole cache, but I think using `"*"` is kind of risky because that's supposed to have a specific meaning which is different. So I'd rather use something like "**"

**Comment by Dieterbe** [comment]
`**` ? why not just support empty pattern then?

**Comment by replay** [comment]
@dieterbe because i think that `""` could easily be specified by accident, that's all :)

**Comment by replay** [comment]
@Dieterbe is there something more that i need to do to make this mergable? (apart from rebasing once again if the cancellable requests get merged)

**Comment by replay** [comment]
@Dieterbe good news, because I'm sure you're looking for more PRs to review, this one would be ready again^^
It also hides nodes away behind an interface in order to make the mockable, that was necessary because I wanted to add a test to test propagation of the cache deletes.
It also makes the chunk cache away of which archives belong to which raw metric. That's necessary because when we delete the metric `xyz` out of the cache we'll also want to delete all the related archives like `xyz_600_sum` or whatever.

File: api/models/ccache.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+func (cd CCacheDelete) Trace(span opentracing.Span) {
+	span.SetTag("patterns", cd.Patterns)
+	span.SetTag("org", cd.OrgId)
+	span.SetTag("propagate", cd.Propagate)
+}
+
+func (cd CCacheDelete) TraceDebug(span opentracing.Span) {
+}
+
+type CCacheDeleteResp struct {
```
Comment (Dieterbe) [inline api/models/ccache.go]: the number of errors is interesting, but much more useful I think would be the first error encountered (for each node)

File: mdata/cache/accnt/lru.go [hunk 3]
```diff
@@ -24,6 +24,10 @@ func (l *LRU) touch(key interface{}) {
 	}
 }
 
+func (l *LRU) del(key interface{}) {
+	delete(l.items, key)
```
Comment (Dieterbe) [inline mdata/cache/accnt/lru.go]: if it's ok for the delete not to affect the `list`, then a comment should explain why

File: mdata/cache/accnt/flat_accnt.go [hunk 4]
```diff
@@ -139,12 +161,19 @@ func (a *FlatAccnt) eventLoop() {
...
+			case evnt_get_total:
+				payload := event.pl.(*GetTotalPayload)
+				a.getTotal(payload.res_chan)
 			case evnt_stop:
 				return
 			case evnt_reset:
 				a.metrics = make(map[string]*FlatAccntMet)
 				a.total = 0
 				a.lru.reset()
+				cacheSizeUsed.SetUint64(a.total)
```
Comment (Dieterbe) [inline mdata/cache/accnt/flat_accnt.go]: it seems throughout all code, cacheSizeUsed is kept in sync with a.total ; so why do we track the same information in two different numbers? can't we just answer reads by looking at cacheSizeUsed?

File: mdata/cache/if.go [hunk 5]
```diff
@@ -7,16 +7,23 @@ import (
 )
 
 type Cache interface {
```
Comment (Dieterbe) [inline mdata/cache/if.go]: could you update this type and add sensible variable names for all parameters. (and also for the methods of the implementations) thanks.

File: mdata/cache/if.go [hunk 6]
```diff
@@ -7,16 +7,23 @@ import (
...
-	Add(string, uint32, chunk.IterGen)
+	Add(string, string, uint32, chunk.IterGen)
 	CacheIfHot(string, uint32, chunk.IterGen)
 	Stop()
 	Search(context.Context, string, uint32, uint32) *CCSearchResult
+	DelMetric(string) *CCDelMetricResult
+	Reset() *CCDelMetricResult
```
Comment (Dieterbe) [inline mdata/cache/if.go]: pointer to CCDelMetricResult is overkill in both cases

File: mdata/cache/ccache.go [hunk 7]
```diff
@@ -94,29 +122,47 @@ func (c *CCache) CacheIfHot(metric string, prev uint32, itergen chunk.IterGen) {
...
+		// if we do not have this raw key yet, create an entry for it
+		ccms, ok := c.metricRawKeys[rawMetric]
+		if !ok {
+			ccms = make(map[string]struct{})
+			c.metricRawKeys[rawMetric] = ccms
+		}
+
+		// if we don't have it yet, associate the metric with this raw key
+		if _, ok = ccms[metric]; !ok {
+			ccms[metric] = struct{}{}
```
Comment (Dieterbe) [inline mdata/cache/ccache.go]: note if we're adding the raw one itself, we don't need to associate it with itself (e.g. `if metric != rawMetric`)

File: mdata/cache/cache_mock.go [hunk 8]
```diff
@@ -12,9 +14,17 @@ type MockCache struct {
 	CacheIfHotCb    func()
 	StopCount       int
 	SearchCount     int
+	DelMetricRes    CCDelMetricResult
+	DelMetricKeys   []string
+}
+
+func NewMockCache() *MockCache {
+	return &MockCache{
+		DelMetricKeys: make([]string, 0),
```
Comment (Dieterbe) [inline mdata/cache/cache_mock.go]: can just leave it at zero value

File: mdata/cache/ccache_test.go [hunk 9]
```diff
@@ -502,3 +502,99 @@ func testSearchDisconnectedWithGapStartEnd(t *testing.T, spanaware, ascending bo
...
+	delRes = cc.Reset()
+	expectDelSeries = 2
+	if delRes.Series != expectDelSeries {
+		t.Fatalf("Expected exactly %d series to get deleted, but got %d", expectDelSeries, delRes.Series)
+	}
+
+	expectDelArchives = 3
+	if delRes.Archives != expectDelArchives {
+		t.Fatalf("Expected exactly %d archives to get deleted, but got %d", expectDelArchives, delRes.Archives)
+	}
```
Comment (Dieterbe) [inline mdata/cache/ccache_test.go]: i would like to see after a reset whether the cache behaves properly (e.g. whether adds and searches and deletes still work). that would be out of scope for TestMetricDelete though (and perhaps it can be implemented by a TestReset test function which calls the other test functions for adds, deletes, etc so it doesn't duplicate that logic)

File: api/ccache.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+func (s *Server) ccacheDeleteRemote(ctx context.Context, req *models.CCacheDelete, peer cluster.Node) models.CCacheDeleteResp {
+	var res models.CCacheDeleteResp
+
+	log.Debug("HTTP metricDelete calling %s/ccache/delete", peer.GetName())
+	buf, err := peer.Post(ctx, "ccacheDeleteRemote", "/ccache/delete", *req)
+	if err != nil {
+		log.Error(4, "HTTP ccacheDelete error querying %s/ccache/delete: %q", peer.GetName(), err)
+		if res.Errors == 0 {
```
Comment (Dieterbe) [inline api/ccache.go]: it's impossible here for res.Errors to be anything else, so no need for the if.

File: api/ccache.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+		}
+		res.Errors++
+		return res
+	}
+
+	err = json.Unmarshal(buf, &res)
+	if err != nil {
+		log.Error(4, "HTTP ccacheDelete error unmarshaling body from %s/ccache/delete: %q", peer.GetName(), err)
+		res.Errors++
+		return res
```
Comment (Dieterbe) [inline api/ccache.go]: line not needed

File: api/ccache.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+			res.FirstError = err.Error()
+		}
+		res.Errors++
+		return res
+	}
+
+	err = json.Unmarshal(buf, &res)
+	if err != nil {
+		log.Error(4, "HTTP ccacheDelete error unmarshaling body from %s/ccache/delete: %q", peer.GetName(), err)
+		res.Errors++
```
Comment (Dieterbe) [inline api/ccache.go]: assigning the error to firstError here would be useful

File: metrictank.go [hunk 13]
```diff
@@ -304,7 +304,7 @@ func main() {
 	/***********************************
 	    Start the ClusterManager
 	***********************************/
-	cluster.Start()
+	cluster.Manager.Start()
```
Comment (Dieterbe) [inline metrictank.go]: why this change?

File: metrictank.go [hunk 14]
```diff
@@ -405,7 +405,7 @@ func main() {
 
 	// Leave the cluster. All other nodes will be notified we have left
 	// and so will stop sending us requests.
-	cluster.Stop()
+	cluster.Manager.Stop()
```
Comment (Dieterbe) [inline metrictank.go]: why this change?

File: api/ccache_test.go [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+func TestMetricDeleteWithErrorInPropagation(t *testing.T) {
+	manager := cluster.InitMock()
+
+	expectedDeletedSeries, expectedDeletedArchives := 0, 0
+
+	// define how many series/archives are getting deleted by peer 0
+	resp := models.CCacheDeleteResp{
+		Peers:           map[string]models.CCacheDeleteResp{"2": {Errors: 1}},
```
Comment (Dieterbe) [inline api/ccache_test.go]: why are we setting this? this response is the response of a peer that we propagated to, right? so that peer should not propagate

File: api/ccache_test.go [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+		t.Fatalf("Expected status code %d, but got %d:\n%+v", expectedCode, res.StatusCode, respParsed)
+	}
+}
+
+func TestMetricDeletePropagation(t *testing.T) {
+	manager := cluster.InitMock()
+
+	expectedDeletedSeries, expectedDeletedArchives := 0, 0
+
+	// define how many series/archives are getting deleted by peer 0
```
Comment (Dieterbe) [inline api/ccache_test.go]: for this comment and the below ones, either the numbers are off or the peer name is incorrect

File: mdata/cache/accnt/lru.go (hunk 3 above)
Comment (replay) [reply mdata/cache/accnt/lru.go]: you're right, it actually should also delete from `.list`

Comment (replay) [reply mdata/cache/accnt/lru.go]: https://github.com/grafana/metrictank/pull/555/commits/3439fb0ab49ba2c2f1e89c37c6d1ef6def3ee841

File: mdata/cache/accnt/flat_accnt.go (hunk 4 above)
Comment (replay) [reply mdata/cache/accnt/flat_accnt.go]: right, that makes sense... actually then we could also get rid of the entire method to get the current total, because we can simply read `cacheSizeUsed.Peek()` which uses an atomic read operation

File: mdata/cache/if.go (hunk 5 above)
Comment (replay) [reply mdata/cache/if.go]: I'm not sure what you mean with `and also for the methods of the implementations`. Are the current ones not sensible enough?
Anyway, I've added the names to the interfaces: https://github.com/grafana/metrictank/pull/555/commits/2b1ad9bd8c14f60b5d27e1aa16d893ebfdce4950

File: mdata/cache/if.go (hunk 6 above)
Comment (replay) [reply mdata/cache/if.go]: https://github.com/grafana/metrictank/pull/555/commits/3abd68cdf4aa473efc5885edea86176ac6e5ac01

File: mdata/cache/ccache.go (hunk 7 above)
Comment (replay) [reply mdata/cache/ccache.go]: I think that would make `Reset()` and `DelMetric()` unnecessarily complicated. 
Also, it would actually make it impossible to return the accurate series count of deleted metrics from `Reset()`, because if the `rawMetric` is not in `metricRawKeys` (f.e. if only the raw metric has been added without any archive) then reset can't reliably know how many series have been cleared, because inside `.metricCache` the only way to determine whether a key is a raw metric key or an archive key would be by pattern matching for the archive suffix like `_sum_<ttl>` (but that would be really ugly).

File: mdata/cache/cache_mock.go (hunk 8 above)
Comment (replay) [reply mdata/cache/cache_mock.go]: ok, fixed https://github.com/grafana/metrictank/pull/555/commits/c90ad3f3c5f7824301122ee313bb0ad05fd4bfe6

File: mdata/cache/ccache_test.go (hunk 9 above)
Comment (replay) [reply mdata/cache/ccache_test.go]: how's that: https://github.com/grafana/metrictank/pull/555/commits/a406e734049abd2daa4cc3422f969b6b936fd58d

File: metrictank.go (hunk 13 above)
Comment (replay) [reply metrictank.go]: That's not necessary anymore, it was necessary in an earlier version of this PR, removing this change.

File: metrictank.go (hunk 14 above)
Comment (replay) [reply metrictank.go]: same ^^

File: api/ccache_test.go (hunk 15 above)
Comment (replay) [reply api/ccache_test.go]: Right, removing that line

File: api/ccache_test.go (hunk 16 above)
Comment (replay) [reply api/ccache_test.go]: moved the peer population into a loop: https://github.com/grafana/metrictank/pull/555/commits/979bea4d3b15129415dfaf202ee9c037955ff604

**Comment by replay** [comment]
@Dieterbe this would be ready to review again

File: mdata/cache/if.go (hunk 5 above)
Comment (Dieterbe) [reply mdata/cache/if.go]: i added a few f545bc416b14dc6a36d16c3cc05a8f234b30d126 but yeah most implementations were already good

File: mdata/cache/ccache.go (hunk 7 above)
Comment (Dieterbe) [reply mdata/cache/ccache.go]: > if the rawMetric is not in metricRawKeys (f.e. if only the raw metric has been added without any archive) then reset can't reliably know how many series have been cleared,

isn't it just a matter of following the associations? if there's 1 key to delete and it has no associations, then it's just that key, no?

File: api/ccache_test.go [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+	buf.ReadFrom(res.Body)
+	json.Unmarshal(buf.Bytes(), &respParsed)
+
+	if len(cache.DelMetricKeys) != 1 || cache.DelMetricKeys[0] != testKey {
+		t.Fatalf("Expected that key %s has been deleted, but it has not", testKey)
+	}
+
+	if respParsed.DeletedSeries != delSeries || respParsed.DeletedArchives != delArchives {
+		t.Fatalf("Expected %d series and %d archives to get deleted, but got %d and %d", delSeries, delArchives, respParsed.DeletedSeries, respParsed.DeletedArchives)
+	}
```
Comment (Dieterbe) [inline api/ccache_test.go]: i'm confused about why we tell the cache to return these fake, made up numbers and then check them, while we also check the real numbers above. can this be simplified ? for example have the mockcache return that deleted series == deleted metric keys, and deleted archives is maybe pinned to 3x the number of deleted keys? does this idea make sense? if not, add a comment somewhere that explains this please

File: cluster/manager.go [hunk 18]
```diff
@@ -141,14 +141,22 @@ func (c *MemberlistManager) setList(list *memberlist.Memberlist) {
 }
 
 func (c *MemberlistManager) ThisNode() Node {
+	return c.thisNode()
+}
+
+func (c *MemberlistManager) thisNode() HTTPNode {
```
Comment (Dieterbe) [inline cluster/manager.go]: why do we need this private function if we already had the public one that did the same?

File: mdata/cache/ccache_test.go (hunk 9 above)
Comment (Dieterbe) [reply mdata/cache/ccache_test.go]: looks good :+1:

File: mdata/cache/ccache_metric.go [hunk 19]
```diff
@@ -19,6 +19,8 @@ type CCacheMetric struct {
 
 	// the list of chunk time stamps in ascending order
 	keys []uint32
+
+	RawMetric string
```
Comment (Dieterbe) [inline mdata/cache/ccache_metric.go]: as i have mentioned before, our codebase is very GC-heavy. we need to pay attention to how many pointers we keep active. and a string has a pointer. instead of keeping the entire RawMetric string for each CCacheMetric why not just keep 1 `uint8` which holds the number to be subtracted from the key to get the raw key? it'll go easier on GC and save us memory as well

**Comment by Dieterbe** [comment]
@replay another round of comments and a few small commits. please have a look :)

File: mdata/cache/ccache_metric.go (hunk 19 above)
Comment (Dieterbe) [reply mdata/cache/ccache_metric.go]: (we could also store the len of the raw string in a uint8 but then we're capped to a key len of 255 which some metrics will want to exceed. but a suffix can't be longer than 255 chars, we know that for sure)

Comment (replay) [reply mdata/cache/ccache_metric.go]: https://github.com/grafana/metrictank/pull/555/commits/b5de5f35733dc491c2842e8f1762a67ad7f98cd7

File: api/ccache_test.go (hunk 17 above)
Comment (replay) [reply api/ccache_test.go]: This is only testing the request handler, not the cache. It's testing one request with propagation disabled (`TestMetricDelete`) and one with propagation enabled (`TestMetricDeletePropagation`).  I'm not sure how creating additional rules like `archives = 3 * series` would simplify things, that would rather just make it more complicated because in order to understand the test a reader would then first need to be aware of this rule.
Maybe I should rename the tests to `TestMetricDeleteRequestHandlerWithoutPropagation()` and `TestMetricDeleteRequestHandlerWithPropagation`?

File: cluster/manager.go (hunk 18 above)
Comment (replay) [reply cluster/manager.go]: right, that's not necessary anymore, was only necessary in some previous version :+1:

Comment (replay) [reply cluster/manager.go]: https://github.com/grafana/metrictank/pull/555/commits/521206c90abb468401cbae77e0fd767e01f76b59

File: api/ccache_test.go (hunk 17 above)
Comment (Dieterbe) [reply api/ccache_test.go]: it's confusing because the mock cache claims to have deleted a number series and archives that has nothing to do with the delete request we actually issued on it. the numbers don't seem to make sense. we assert that it only received 1 metric key delete , then how could any cache have deleted 3 series if there was only 1 key? i know it's a mock and not a real cache, but the numbers should make more sense I think.

Comment (replay) [reply api/ccache_test.go]: ah, ok i understand what you mean

Comment (replay) [reply api/ccache_test.go]: how's that https://github.com/grafana/metrictank/pull/555/commits/58d279a619d62cdac74b1f286436f19d0356d693

Comment (Dieterbe) [reply api/ccache_test.go]: that would clear things up if the DelMetric method took patterns.
but according to the argument names or the interface function and its implementations
(and the docs for CCache.DelMetric) it only takes 1 metric key, not a pattern.

Comment (replay) [reply api/ccache_test.go]: i found a bug while looking at that just now: https://github.com/grafana/metrictank/pull/555/commits/b168558e98f56d0dc7d4aaca2e99772cd6c7b3b4
